### DISCUSSION TITLE
`splitAtom` does not work with `async` atoms

### DISCUSSION THREAD

**Discussion by dlindahl** [pmndrs/jotai#1751] [Q&A] [answered]
I'm upgrading my application to the new v2 API and ran into an issue where I'm using `splitAtom` with an `async` atom. This worked previously but now returns the error `arr.forEach is not a function` because the atom returns a `Promise` rather than an array of resolved values. This is important in my application because accessing the `async` atom or the `splitAtom` currently results in a suspended state until the `async` item resolves.

Is this an intentional design decision or was `splitAtom` forgotten about during the change in `read` behavior? If its an intentional design decision, is there a preferred method of integrating `splitAtom` with `async` atoms? I'd also recommend a note in the docs or the migration guide about any issue here.

Thanks!

Reproduction:
https://codesandbox.io/s/react-typescript-forked-zoyqkd?file=/src/App.tsx

**Comment by dai-shi** [comment]
Thanks for reporting. I think I forgot... Didn't notice as missing tests. It should probably fixable. It would require a big refactor. Let me try.

**Reply by dai-shi** [reply]
Hm, that is way too complicated.
Also noticed that the example doesn't fully work in v1 anyway: https://codesandbox.io/s/react-typescript-forked-chx282?file=/src/App.tsx
So, it's a half regression in behavior, but it's by design, technically since v1.

**Comment by dai-shi** [comment, marked as answer]
Here's one solution in v2.
https://codesandbox.io/s/react-typescript-forked-y94yfw
It's using an unstable util. We could use loadable too.
I'll document it later.

Another solution is atoms-in-atom. We are still experimenting but seems promising.
Related discussion: https://github.com/jotaijs/jotai-tanstack-query/issues/21 & https://github.com/pmndrs/jotai/discussions/1615

**Reply by dai-shi** [reply]
#1755

**Reply by dlindahl** [reply]
This is essentially what I wound up doing and its good to see this mentioned in the docs. Thanks!

**Comment by karthikbadam** [comment]
> Here's one solution in v2.
> https://codesandbox.io/s/react-typescript-forked-y94yfw

I am seeing an error when I `edit` or `remove` a todo item in this solution.
```
Error
not writable atom
...
> 34 |     setTodo((oldValue) => ({ ...oldValue, task: e.target.value }));
```

Is there a workaround for it?

**Reply by dai-shi** [reply]
Right, but it wasn't possible with v1 either: https://codesandbox.io/s/react-typescript-forked-jdw4og?file=/src/App.tsx

Would you like to edit the example to make it similar to your use case?

**Reply by karthikbadam** [reply]
Thanks! I see. I started using jotai with v2, so I was not sure. 

I have a similar use case as the todo example:
- an async atom containing a collection of "items" fetched from an API
- a component that renders each "item" and supports `remove/modify/add` operations

I was considering using splitAtom.
I think I can use [action atoms](https://jotai.org/docs/guides/composing-atoms#action-atoms) instead to update the base atom. Are there any alternative ways?

**Reply by dai-shi** [reply]
> I started using jotai with v2, so I was not sure.

Ooops, my misunderstanding. `splitAtom` must receive a "writable" atom to return each atom writable/modifiable. You can't simply modify read-only atom, right?
If you fetch items from an API, it will be read-only atom.
`atomWithDefault` is probably something to help for your case.

**Reply by karthikbadam** [reply]
That makes sense. Thanks!
Switching to `atomWithDefault` works in the todo example.
https://codesandbox.io/s/react-typescript-forked-pwc2t1?file=/src/App.tsx

I will try it out in my use case too.
